### ISSUE TITLE
Pydantic not recognized on extra search paths

### ISSUE THREAD

**Issue by JamieForsythe** [astral-sh/ty#4159]
### Summary

When instantiating a pydantic model with `extra="allow"` and passing attributes that aren't explicitly defined on the model, ty seems to display inconsistent behaviour.

- On macOS Tahoe 26.6 machine and Python 3.11, `ty check` produces an unknown-argument error
- On macOS Tahoe 26.6 machine and Python 3.12/3.13, `ty check` runs successfully without error
- On Ubuntu 24.04.4 GitHub Actions runner and Python 3.11, `ty check` also runs successfully without error

Annoyingly, this means that a `ty: ignore` comment is needed on Python 3.11 macOS, but in the other contexts this produces unused-ignore-comment warnings/errors.

Minimal repro code:

```python
from pydantic import BaseModel, ConfigDict


class MyModel(BaseModel):
    model_config = ConfigDict(extra="allow")


foo = MyModel(a=1)
```

Have also created a repro repo [here](https://github.com/JamieForsythe/ty-inconsistency-repro)

I haven't been able to reproduce this behaviour in a non-pydantic context, so I'm not sure whether the issue is limited to that or is broader. I don't really mind whether ty accommodates Pydantic models with non-explicit attributes or not, as I appreciate it's potentially beyond the scope of ty, it's just the inconsistency in ty's behaviour that's annoying and causes issues in our CI pipelines.

Output from `ty check -vv src/` on macOS Tahoe 26.6, Python 3.11:

```text
2026-08-03 09:42:47.996022 DEBUG Version: 0.0.63 (46f4915e6 2026-07-23)
2026-08-03 09:42:48.011528 DEBUG Architecture: aarch64, OS: macos
2026-08-03 09:42:48.011541 DEBUG Searching for a project in '/Users/jamieforsythe/dev/personal/ty-inconsistency-repro'
2026-08-03 09:42:48.01166 DEBUG Resolving requires-python constraint: `>=3.11`
2026-08-03 09:42:48.011682 DEBUG Resolved requires-python constraint to: 3.11
2026-08-03 09:42:48.011688 DEBUG Found project at '/Users/jamieforsythe/dev/personal/ty-inconsistency-repro'
2026-08-03 09:42:48.011692 DEBUG Searching for a user-level configuration at `/Users/jamieforsythe/.config/ty/ty.toml`
2026-08-03 09:42:48.015498 INFO Defaulting to python-platform `darwin`
2026-08-03 09:42:48.015522 DEBUG Resolving `VIRTUAL_ENV` environment variable: /Users/jamieforsythe/dev/personal/ty-inconsistency-repro/.venv
2026-08-03 09:42:48.01571 DEBUG Attempting to parse virtual environment metadata at '/Users/jamieforsythe/dev/personal/ty-inconsistency-repro/.venv/pyvenv.cfg'
2026-08-03 09:42:48.015792 DEBUG Searching for site-packages directory in sys.prefix /Users/jamieforsythe/dev/personal/ty-inconsistency-repro/.venv
2026-08-03 09:42:48.015816 DEBUG Resolved site-packages directories for this virtual environment are: ["/Users/jamieforsythe/dev/personal/ty-inconsistency-repro/.venv/lib/python3.11/site-packages"]
2026-08-03 09:42:48.01589 DEBUG Searching for real stdlib directory in sys.prefix /Users/jamieforsythe/.local/share/uv/python/cpython-3.11.15-macos-aarch64-none
2026-08-03 09:42:48.015898 DEBUG Resolved real stdlib path for this virtual environment is: /Users/jamieforsythe/.local/share/uv/python/cpython-3.11.15-macos-aarch64-none/lib/python3.11
2026-08-03 09:42:48.015998 DEBUG Including `./src` in `environment.root` because a `./src` directory exists and is not a package
2026-08-03 09:42:48.01601 DEBUG Adding `/Users/jamieforsythe/dev/personal/ty-inconsistency-repro` from the `PYTHONPATH` environment variable to `extra_paths`
2026-08-03 09:42:48.016015 DEBUG Adding `/Users/jamieforsythe/dev/personal/ty-inconsistency-repro/src` from the `PYTHONPATH` environment variable to `extra_paths`
2026-08-03 09:42:48.016019 DEBUG Adding `/Users/jamieforsythe/dev/personal/ty-inconsistency-repro/.venv/lib/python3.11/site-packages` from the `PYTHONPATH` environment variable to `extra_paths`
2026-08-03 09:42:48.016032 DEBUG Adding extra search-path `/Users/jamieforsythe/dev/personal/ty-inconsistency-repro`
2026-08-03 09:42:48.016042 DEBUG Adding extra search-path `/Users/jamieforsythe/dev/personal/ty-inconsistency-repro/src`
2026-08-03 09:42:48.016056 DEBUG Adding extra search-path `/Users/jamieforsythe/dev/personal/ty-inconsistency-repro/.venv/lib/python3.11/site-packages`
2026-08-03 09:42:48.016085 DEBUG Adding first-party search path `/Users/jamieforsythe/dev/personal/ty-inconsistency-repro/src`
2026-08-03 09:42:48.016089 DEBUG Adding first-party search path `/Users/jamieforsythe/dev/personal/ty-inconsistency-repro`
2026-08-03 09:42:48.016093 DEBUG Using vendored stdlib
2026-08-03 09:42:48.016206 DEBUG Adding site-packages search path `/Users/jamieforsythe/dev/personal/ty-inconsistency-repro/.venv/lib/python3.11/site-packages`
2026-08-03 09:42:48.016217 INFO Python version: Python 3.11, platform: darwin
2026-08-03 09:42:48.016224 DEBUG Adding new file root '/Users/jamieforsythe/dev/personal/ty-inconsistency-repro' of kind Project
2026-08-03 09:42:48.016245 DEBUG Adding new file root '/Users/jamieforsythe/dev/personal/ty-inconsistency-repro/src' of kind SearchPath
2026-08-03 09:42:48.016274 DEBUG Adding new file root '/Users/jamieforsythe/dev/personal/ty-inconsistency-repro/.venv/lib/python3.11/site-packages' of kind SearchPath
2026-08-03 09:42:48.016809 DEBUG Setting included paths: 1
2026-08-03 09:42:48.017719 DEBUG Reloading files for project `ty-inconsistency-repro`
2026-08-03 09:42:48.018251 DEBUG Starting main loop
2026-08-03 09:42:48.018301 DEBUG Waiting for next main loop message.
2026-08-03 09:42:48.018487 DEBUG Checking all files in project 'ty-inconsistency-repro'
2026-08-03 09:42:48.022147 INFO Indexed 1 file(s) in 0.004s
2026-08-03 09:42:48.022573 DEBUG Checking file '/Users/jamieforsythe/dev/personal/ty-inconsistency-repro/src/ty_inconsistency_repro/repro.py'
2026-08-03 09:42:48.023024 DEBUG Resolving dynamic module resolution paths
2026-08-03 09:42:48.046418 DEBUG Module `__builtins__` not found in search paths
2026-08-03 09:42:48.083214 DEBUG Checking all files took 0.061s
error[unknown-argument]: Argument `a` does not match any known parameter
 --> src/ty_inconsistency_repro/repro.py:8:15
  |
8 | foo = MyModel(a=1)
  |               ^^^
  |
info: rule `unknown-argument` is enabled by default

Found 1 diagnostic
2026-08-03 09:42:48.083311 DEBUG Exiting main loop
```

### Version

ty 0.0.63 (46f4915e6 2026-07-23)

**Comment by sharkdp** [comment]
Thank you for reporting this. In your logs, you have:

> 2026-08-03 09:42:48.016019 DEBUG Adding `/Users/jamieforsythe/dev/personal/ty-inconsistency-repro/.venv/lib/python3.11/site-packages` from the `PYTHONPATH` environment variable to `extra_paths`

which means that your `.venv` is added to extra-search-paths. However, special support for Pydantic is only activated when `pydantic` is imported from a third-party directory and we currently don't consider extra search paths as definitely containing 3rd party code. Even if your setup is maybe a bit unusual, I think we should support this. I'm working on a fix.

**Comment by JamieForsythe** [comment]
Ahhh ok, wasn't aware of that behaviour. I've just removed my venv from my PYTHONPATH variable and confirmed that it does fix the issue.

Thanks for the speedy response!

**Comment by sharkdp** [comment]
Thank you for the excellent bug report. Including the logs was key 😄
